spec/integration/helper.rb

```diff
@@ -0,0 +1,452 @@
+  # Docker::Error::ServerError if that's not the case.
+  container.top
+  container
+rescue Docker::Error::ServerError
+  container.delete(force: true)
+  nil
+rescue Docker::Error::NotFoundError
+  nil
+end
+
+# Ping the mysql container within the portus container. It will raise an error
+# if a ping attempt fails 5 times.
+def mysql_ping!
+  conn = "--count=1 --connect-timeout=1"
+  cred = "--host=integration_db -u root --password=portus"
+
+  success = docker_exec("integration_portus", "mysqladmin #{conn} #{cred} ping", 5)
+  raise StandardError, "Could not ping!" unless success
+end
+
+# Setup the database container.
+def setup_db!
+  start_container!(
+    "library/mariadb:10.0.23",
+    "integration_db",
+    "Env" => ["MYSQL_ROOT_PASSWORD=portus"]
+  )
+end
+
+# Setup the Portus container.
+#   - rebuild: whether the image has to be rebuilt or not.
+#   - env: a list of additional environment variables to be passed.
+#   - ip: the IP that the Portus container should have.
+def setup_portus!(rebuild = true, env = [], address = nil)
+  dir = File.expand_path(File.dirname(__FILE__) + "../../../")
+
+  # Build Portus with the current code and run it in a container.
+  Dir.chdir(dir) do
+    if rebuild
+      # Docker::Image.build_from_dir fails spectacularly on this. Just execute
+      # the damn command.
+      puts "Building Portus from directory: #{dir}"
+      PTY.spawn("docker build -t integration_portus:latest .") do |stdout, _, _|
+        # rubocop:disable Lint/HandleExceptions
+        begin
+          stdout.each { |line| print line }
+        rescue Errno::EIO
+          # End of output
+        end
+        # rubocop:enable Lint/HandleExceptions
+      end
+    end
+
+    opts = {
+      "Env"          => [
+        "PORTUS_MACHINE_FQDN_VALUE=#{ip}",
+        "PORTUS_DB_HOST=integration_db",
+        "RAILS_ENV=test",
+        env
+      ].flatten,
+      "Volumes"      => { Dir.pwd => { "/portus" => "" } },
+      "Cmd"          => "puma -b tcp://0.0.0.0:3000 -w 10".split(" "),
+      "ExposedPorts" => { "3000/tcp" => {} },
+      "HostConfig"   => {
+        "Binds"        => ["#{Dir.pwd}:/portus"],
+        "Links"        => ["integration_db"],
+        "PortBindings" => { "3000/tcp" => [{ "HostIP" => "0.0.0.0", "HostPort" => "3000" }] }
+      }
+    }
+    opts["NetworkSettings"] = { "IPAddress" => address } unless address.nil?
+    start_container!("integration_portus:latest", "integration_portus", opts)
+  end
+end
+
+def setup_templates!
+  # Render the template
+  @ip = ip
+  src = File.expand_path(File.dirname(__FILE__)) + "/" + ("fixtures")
+  tpl = File.read(File.join(src, "config.yml.erb"))
+  res = ERB.new(tpl, nil, "<>").result(binding)
+
+  # Write it
+  output = File.join(src, "config.yml")
+  File.open(output, "w") { |file| file.write(res) }
+end
+
+# Get the IP from docker0
+def ip
+  docker0 = `/sbin/ifconfig docker0`
+  ips     = docker0.scan(/(([0-9]{1,3}[\.]){3}[0-9]{1,3})/)
+  ips.first.first
+end
```

Comment:
We can maybe do this when we run integration tests inside of a vm to test multiple docker versions ? I'm saying this because I want to mess with the host the least as possible. For example, I have a registry listening in :5000 that is enabled, so this port will be busy and tests will fail when trying to start the registry container.